qiskit_experiments/visualization/style.py

```diff
@@ -0,0 +1,118 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Configurable stylesheet for :class:`BasePlotter` and :class:`BaseDrawer`.
+"""
+from copy import copy
+from typing import Dict, Tuple
+
+from qiskit_experiments.framework import Options
+
+
+class PlotStyle(Options):
```

Comment:
Each `BaseDrawer` subclass must define a default style, which should inherit from the parent drawer. This allows users to only define style-parameters they want to overwrite. The default style of a drawer subclass is defined in `_default_style()`, allowing subclasses to add extra style parameters easily. This is a similar mechanism to `options`, `plot_options`, and `experiment_options`. To be consistent with other `_default_*` functions, the style type would have to have an interface similar to `Options`. To combine the default style with the custom style, defined by a user, the style type also needs to allow merging and updating; exactly like a dictionary. It was easier to extend `Options` --- and add `merge()` and `update()` --- than subclass `dict` and add functionality equivalent to that provided by `Options`. This inheritance also means `PlotStyle` benefits from value validation.

Reply:
I completely agree that nested `Options` are actually bad. I checked, and `Options` doesn't support "namespaced" options being set unless it's as a dictionary key (`opts["key-name"] = ...`), which limits setting them using ``plotter.options.style.foo1 = "boo"``. Having a dedicated class --- instead of merging styles with `options` --- also allows `BaseDrawer` to separate the default style parameters from custom style parameters, so that users can view only those options which have changed. This is the reason I added `default_style` and `custom_style` options. The same functionality could be achieved using`options` and a set of altered options, but then it wouldn't be clear which options are style parameters and which aren't as the names may not be clear; unless we use namespaced names (which is difficult with `Options`).

Given other comments also bring into question whether `PlotStyle` should subclass `Options`, I have refactored the `PlotStyle` class to inherit from `dict` (see 7c3ca90 and 8ed2c9f). I have kept style parameters separate from `options` and `figure_options` be keeping the `default_style` and `custom_style` options. This allows `BaseDrawer` subclasses to merge and update style parameters (see `BaseDrawer.style`) while keeping them in a manageable container. Regarding namespacing style parameters, the fullstop in variable names makes it difficult to use PlotStyle. For example, style parameters cannot be set in `__init__` as kwargs (i.e., `PlotStyle(tick.text_size=14)` doesn't work). Because of this, I've left the style parameter names with underscores instead of fullstops. I do think the namespaced/grouped interface would be useful, and should be considered as the visualization module is extended.